crates/python/src/api.rs

```diff
@@ -0,0 +1,223 @@
+    },
+    qpu::{
+        quilc::{CompilerOpts, TargetDevice, DEFAULT_COMPILER_TIMEOUT},
+        Qcs,
+    },
+};
+use rigetti_pyo3::{
+    create_init_submodule, py_wrap_data_struct, py_wrap_error, py_wrap_type, py_wrap_union_enum,
+    wrap_error, ToPython,
+};
+
+create_init_submodule! {
+    classes: [
+        PyExecutionResult,
+        PyExecutionResults,
+        PyRegister,
+        PyRewriteArithmeticResult,
+        PyTranslationResult
+    ],
+    errors: [
+        InvalidConfigError,
+        ExecutionError,
+        TranslationError,
+        CompilationError,
+        RewriteArithmeticError,
+        DeviceIsaError,
+        QcsSubmitError
+    ],
+    funcs: [
+        compile,
+        rewrite_arithmetic,
+        translate,
+        submit,
+        retrieve_results,
+        build_patch_values,
+        get_quilc_version
+    ],
+}
+
+py_wrap_data_struct! {
+    PyRewriteArithmeticResult(RewriteArithmeticResult) as "RewriteArithmeticResult" {
+        program: String => Py<PyString>,
+        recalculation_table: Vec<String> => Py<PyList>
+    }
+}
+
+py_wrap_data_struct! {
+    PyTranslationResult(TranslationResult) as "TranslationResult" {
+        program: String => Py<PyString>,
+        ro_sources: Option<HashMap<String, String>> => Option<Py<PyDict>>
+    }
+}
+
+py_wrap_type! {
+    PyExecutionResults(ExecutionResults) as "ExecutionResults";
+}
+
+py_wrap_union_enum! {
+    PyRegister(Register) as "Register" {
+        f64: F64 => Vec<Py<PyFloat>>,
+        i16: I16 => Vec<Py<PyInt>>,
+        i32: I32 => Vec<Py<PyInt>>,
+        i8: I8 => Vec<Py<PyInt>>,
+        complex64: Complex64 => Vec<Py<PyComplex>>
+    }
+}
+
+py_wrap_type! {
+    PyExecutionResult(ExecutionResult) as "ExecutionResult";
+}
+
+create_exception!(qcs, InvalidConfigError, PyRuntimeError);
+create_exception!(qcs, ExecutionError, PyRuntimeError);
+create_exception!(qcs, TranslationError, PyRuntimeError);
+create_exception!(qcs, CompilationError, PyRuntimeError);
+create_exception!(qcs, RewriteArithmeticError, PyRuntimeError);
+create_exception!(qcs, DeviceIsaError, PyValueError);
+
+wrap_error!(SubmitError(qcs::api::SubmitError));
+py_wrap_error!(api, SubmitError, QcsSubmitError, PyRuntimeError);
+
+#[pyfunction(kwds = "**")]
+pub fn compile<'a>(
+    py: Python<'a>,
+    quil: String,
+    target_device: String,
+    kwds: Option<&PyDict>,
+) -> PyResult<&'a PyAny> {
+    let target_device: TargetDevice =
+        serde_json::from_str(&target_device).map_err(|e| DeviceIsaError::new_err(e.to_string()))?;
+
+    let mut compiler_timeout = Some(DEFAULT_COMPILER_TIMEOUT);
+    if let Some(kwargs) = kwds {
+        if let Some(timeout_arg) = kwargs.get_item("timeout") {
+            let timeout: Result<Option<u8>, _> = timeout_arg.extract();
+            if let Ok(option) = timeout {
+                compiler_timeout = option
+            }
+        }
+    }
+
+    pyo3_asyncio::tokio::future_into_py(py, async move {
+        let client = Qcs::load()
```

Comment:
Do we want / need to support passing this in as a parameter? If so, what's the right pattern for reconstructing it from a (guessing here) `PyDict` or `PyAny`?

Reply:
Shouldn't we follow the same pattern as with the rest of the Rust-defined structs and just let Python users manipulate something that is otherwise kept on the Rust side? If the RUst SDK is making all the network calls, it makes sense to me to stick with that as the canonical ClientConfiguration rather than using the one from the `qcs-api-client-python` side. That Python crate shouldn't even need to be imported by `pyquil` once we're done with v4.

Reply:
Fix is in https://github.com/rigetti/qcs-sdk-rust/pull/235